File: app/android/app/src/main/java/com/nativestream/android/domain/model/control/Envelope.kt
// app/src/main/java/com/nativestream/android/domain/model/control/Envelope.kt
//
// LMC protocol types — mirrors server/control/protocol.go.
// All WebSocket messages between Android and the Go broker use these types.

package com.nativestream.android.domain.model.control

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

@Serializable
enum class MessageType {
    @SerialName("register")     REGISTER,
    @SerialName("session_list") SESSION_LIST,
    @SerialName("play")         PLAY,
    @SerialName("stop")         STOP,
    @SerialName("pull_back")    PULL_BACK,
    @SerialName("pull_back_ack") PULL_BACK_ACK,
    @SerialName("state_update") STATE_UPDATE,
    @SerialName("volume_set")   VOLUME_SET,
    @SerialName("ping")         PING,
    @SerialName("pong")         PONG,
}

@Serializable
enum class DeviceKind {
    @SerialName("controller") CONTROLLER,
    @SerialName("target")     TARGET,
    @SerialName("tv")         TV,
}

@Serializable
data class Envelope(
    val type:    MessageType,
    val from:    String,
    val to:      String,
    val auth:    String? = null,
    // JsonElement round-trips as a real nested JSON value (object/array/etc.)
    // matching Go's json.RawMessage on the wire, instead of being
    // double-encoded as a JSON string.
    val payload: JsonElement = JsonObject(emptyMap()),
)

@Serializable
data class SessionInfo(
    @SerialName("device_id")    val deviceId:    String,
    val name:                                    String,
    val kind:                                    DeviceKind,
    @SerialName("channel_id")   val channelId:   String,
    @SerialName("channel_name") val channelName: String,
    @SerialName("stream_url")   val streamUrl:   String,
    val playing:                                 Boolean,
    @SerialName("connected_at") val connectedAt: String,
)

// ── Payload types ─────────────────────────────────────────────────────────────

@Serializable
data class RegisterPayload(
    val name: String,
    val kind: DeviceKind,
)

@Serializable
data class PlayPayload(
    @SerialName("channel_id")   val channelId: String,
    @SerialName("channel_name") val channelName: String,
    @SerialName("stream_url")   val streamUrl: String,
)

@Serializable
data class PullBackPayload(
    @SerialName("from_device") val fromDevice: String,
)

@Serializable
data class PullBackAckPayload(
    @SerialName("channel_id")   val channelId: String,
    @SerialName("channel_name") val channelName: String,
    @SerialName("stream_url")   val streamUrl: String,
)

@Serializable
data class StateUpdatePayload(
    @SerialName("channel_id")   val channelId: String,
    @SerialName("channel_name") val channelName: String,
    @SerialName("stream_url")   val streamUrl: String,
    val playing: Boolean,
)

@Serializable
data class VolumeSetPayload(
    val level: Float, // 0.0 – 1.0
)

@Serializable
data class SessionListPayload(
    val sessions: List<SessionInfo>,
)

// ── Envelope helpers ─────────────────────────────────────────────────────────

/**
 * Builds an Envelope with [payload] encoded into a real JsonElement so it
 * matches Go's json.RawMessage on the wire instead of being stringified.
 */
inline fun <reified T> buildEnvelope(
    type: MessageType,
    from: String,
    to: String,
    payload: T,
    auth: String? = null,
): Envelope {
    val element = kotlinx.serialization.json.Json.encodeToJsonElement(
        kotlinx.serialization.serializer<T>(), payload
    )
    return Envelope(type = type, from = from, to = to, auth = auth, payload = element)
}

/**
 * Decodes the Envelope's payload into [T].
 */
inline fun <reified T> Envelope.decodePayload(): T? =
    try {
        kotlinx.serialization.json.Json.decodeFromJsonElement(
            kotlinx.serialization.serializer<T>(), payload
        )
    } catch (e: Exception) {
        null
    }

File: app/macos/NativeStream/NativeStream/Protocol/Envelope.swift
/// Protocol/Envelope.swift
///
/// LMC message protocol types — mirrors server/control/protocol.go.
/// All WebSocket messages between macOS and the Go broker use these types.

import Foundation

enum MessageType: String, Codable {
    case register
    case sessionList    = "session_list"
    case play
    case stop
    case pullBack        = "pull_back"
    case pullBackAck      = "pull_back_ack"
    case stateUpdate     = "state_update"
    case ping
    case pong
}

enum DeviceKind: String, Codable {
    case controller
    case target
    case tv // reserved for future TV client
}

enum JSONValue: Codable {
    case object([String: JSONValue])
    case array([JSONValue])
    case string(String)
    case number(Double)
    case bool(Bool)
    case null

    init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if let v = try? c.decode([String: JSONValue].self) { self = .object(v) }
        else if let v = try? c.decode([JSONValue].self) { self = .array(v) }
        else if let v = try? c.decode(String.self) { self = .string(v) }
        else if let v = try? c.decode(Double.self) { self = .number(v) }
        else if let v = try? c.decode(Bool.self) { self = .bool(v) }
        else { self = .null }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch self {
        case .object(let v): try c.encode(v)
        case .array(let v): try c.encode(v)
        case .string(let v): try c.encode(v)
        case .number(let v): try c.encode(v)
        case .bool(let v): try c.encode(v)
        case .null: try c.encodeNil()
        }
    }
}

struct Envelope: Codable {
    let type: MessageType
    let from: String
    let to: String
    let auth: String?
    let payload: JSONValue

    init(type: MessageType, from: String, to: String, auth: String? = nil, payload: JSONValue = .object([:])) {
        self.type = type
        self.from = from
        self.to = to
        self.auth = auth
        self.payload = payload
    }
}

struct SessionInfo: Codable, Identifiable {
    var id: String { deviceID }
    let deviceID: String
    let name: String
    let kind: DeviceKind
    let channelID: String
    let channelName: String
    let streamURL: String
    let playing: Bool
    let connectedAt: String

    enum CodingKeys: String, CodingKey {
        case deviceID    = "device_id"
        case name, kind
        case channelID   = "channel_id"
        case channelName = "channel_name"
        case streamURL   = "stream_url"
        case playing
        case connectedAt = "connected_at"
    }
}

// MARK: - Payload types

struct RegisterPayload: Codable {
    let name: String
    let kind: DeviceKind
}

struct PlayPayload: Codable {
    let channelID: String
    let channelName: String
    let streamURL: String

    enum CodingKeys: String, CodingKey {
        case channelID   = "channel_id"
        case channelName = "channel_name"
        case streamURL   = "stream_url"
    }
}
struct PullBackPayload: Codable {
    let fromDevice: String

    enum CodingKeys: String, CodingKey {
        case fromDevice = "from_device"
    }
}

struct PullBackAckPayload: Codable {
    let channelID: String
    let channelName: String
    let streamURL: String

    enum CodingKeys: String, CodingKey {
        case channelID   = "channel_id"
        case channelName = "channel_name"
        case streamURL   = "stream_url"
    }
}

struct StateUpdatePayload: Codable {
    let channelID: String
    let channelName: String
    let streamURL: String
    let playing: Bool

    enum CodingKeys: String, CodingKey {
        case channelID   = "channel_id"
        case channelName = "channel_name"
        case streamURL   = "stream_url"
        case playing
    }
}

struct SessionListPayload: Codable {
    let sessions: [SessionInfo]
}

// MARK: - Envelope helpers

extension Envelope {
    static func encoding<T: Encodable>(
        type: MessageType, from: String, to: String, payload: T
    ) -> Envelope? {
        guard let data = try? JSONEncoder().encode(payload),
              let value = try? JSONDecoder().decode(JSONValue.self, from: data)
        else { return nil }
        return Envelope(type: type, from: from, to: to, payload: value)
    }

    func decoding<T: Decodable>(as type: T.Type) -> T? {
        guard let data = try? JSONEncoder().encode(payload) else { return nil }
        return try? JSONDecoder().decode(T.self, from: data)
    }
}
